contracts/PodsUpkeep.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.7.6;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+
+import "@pooltogether/pooltogether-generic-registry/contracts/AddressRegistry.sol";
+
+import "./interfaces/IPod.sol";
+import "./interfaces/KeeperCompatibleInterface.sol";
+
+
+///@notice Contract implements Chainlink's Upkeep system interface, automating the upkeep of a registry of Pod contracts
+contract PodsUpkeep is KeeperCompatibleInterface, Ownable {
+
+    using SafeMathUpgradeable for uint256;
+    
+    /// @notice Address of the registry of pods contract which require upkeep
+    AddressRegistry public podsRegistry;
+
+    /// @notice Interval at which pods.batch() will be called
+    uint256 public upkeepBlockInterval;    
+
+    /// @notice Block number when batch was last called successfully
+    uint256 public lastUpkeepBlockNumber;
+
+    /// @notice Minimum pod float fraction
+    uint256 public targetFloatFraction;
+
+    /// @notice Emitted when the upkeep block interval is updated
+    event UpkeepBlockIntervalUpdated(uint upkeepBlockInterval);
+        
+    /// @notice Emitted when the float fraction is updated
+    event TargetFloatFractionUpdated(uint targetFloatFraction);
+
+
+    /// @notice Contract Constructor. No initializer. 
+    constructor(AddressRegistry _podsRegistry, address _owner) Ownable() {
+        
+        podsRegistry = _podsRegistry;
+        
+        transferOwnership(_owner);
+
+        lastUpkeepBlockNumber = block.number;
+
+        // initialize values for block interval and float rate?
+    }
+
+    /// @notice Checks if Pods require upkeep. Call in a static manner every block by the Chainlink Upkeep network.
+    /// @param checkData Not used in this implementation.
+    /// @return upkeepNeeded as true if performUpkeep() needs to be called, false otherwise. performData returned empty. 
+    function checkUpkeep(bytes calldata checkData) override external view returns (bool upkeepNeeded, bytes memory performData) {
+        
+        address[] memory pods = podsRegistry.getAddresses();
+        for(uint256 i = 0; i < pods.length; i++){ // can get out of gas here -- do we want to implement a batchLimit ?
+            if(checkUpkeepRequired(IPod(pods[i]))){
+                return (true, "");
+            }
+        }
+        return (false, "");    
+    }
+
+    /// @notice Performs upkeep on the pods contract
+    /// @param performData Not used in this implementation.
+    function performUpkeep(bytes calldata performData) override external {
+    
+        address[] memory pods = podsRegistry.getAddresses();
+        for(uint256 i = 0; i < pods.length; i++){
+            (bool required, uint256 batchAmount) = amountAboveFloat(IPod(pods[i])); 
+            if(required) {
+                require(IPod(pods[i]).batch(batchAmount), "PodsUpkeep: batch() failed");
```

Comment:
Do we want to require here? A failing batch will prevent the remainder of the pods to not have batch() called on them..